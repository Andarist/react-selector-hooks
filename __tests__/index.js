import * as React from 'react'
import TestRenderer from 'react-test-renderer'
import {
  createSelector,
  createStateSelector,
  createStructuredSelector,
} from '../src'

const render = element => {
  const renderer = TestRenderer.create(element)
  return {
    rerender(element) {
      renderer.update(element)
    },
    snapshot() {
      expect(renderer.toJSON()).toMatchSnapshot()
    },
  }
}

test('createSelector', () => {
  const selector = jest.fn((a, b) => a + b)
  const useSelector = createSelector(selector)

  function Test({ a, b }) {
    const result = useSelector(a, b)
    return <span>{result}</span>
  }

  const { rerender, snapshot } = render(<Test a={2} b={3} />)
  expect(selector).toBeCalledTimes(1)
  expect(selector).toHaveBeenCalledWith(2, 3)
  expect(selector).toHaveReturnedWith(5)
  snapshot()

  rerender(<Test a={2} b={3} />)
  expect(selector).toBeCalledTimes(1)
  snapshot()

  rerender(<Test a={4} b={3} />)
  expect(selector).toBeCalledTimes(2)
  expect(selector).toHaveReturnedWith(7)
  snapshot()
})

test('createStateSelector', () => {
  const selector = jest.fn((a, b) => a + b)
  const useSelector = createStateSelector(
    [(state, path) => state[path], state => state.b],
    selector,
  )

  function Test({ state, path }) {
    const result = useSelector(state, path)
    return <span>{result}</span>
  }

  const obj = { a: 2, b: 3, c: 10 }
  const { rerender, snapshot } = render(<Test path="a" state={obj} />)
  expect(selector).toBeCalledTimes(1)
  expect(selector).toHaveBeenCalledWith(2, 3)
  expect(selector).toHaveReturnedWith(5)
  snapshot()

  rerender(<Test path="a" state={obj} />)
  expect(selector).toBeCalledTimes(1)
  snapshot()

  rerender(<Test path="c" state={obj} />)
  expect(selector).toBeCalledTimes(2)
  expect(selector).toHaveBeenCalledWith(10, 3)
  expect(selector).toHaveReturnedWith(13)
  snapshot()

  const obj2 = { ...obj }
  rerender(<Test path="c" state={obj2} />)
  expect(selector).toBeCalledTimes(2)
  snapshot()

  const obj3 = { ...obj2, b: 6 }
  rerender(<Test path="c" state={obj3} />)
  expect(selector).toBeCalledTimes(3)
  expect(selector).toHaveBeenCalledWith(10, 6)
  expect(selector).toHaveReturnedWith(16)
  snapshot()
})

test('createStructuredSelector', () => {
  const selector = jest.fn(({ x, y }) => x + y)
  const useSelector = createStructuredSelector(
    {
      x: (state, path) => state[path],
      y: state => state.b,
    },
    selector,
  )

  function Test({ state, path }) {
    const result = useSelector(state, path)
    return <span>{result}</span>
  }

  const obj = { a: 2, b: 3, c: 10 }
  const { rerender, snapshot } = render(<Test path="a" state={obj} />)
  expect(selector).toBeCalledTimes(1)
  expect(selector.mock.calls[0][0]).toEqual({ x: 2, y: 3 })
  expect(selector).toHaveReturnedWith(5)
  snapshot()

  rerender(<Test path="a" state={obj} />)
  expect(selector).toBeCalledTimes(1)
  snapshot()

  rerender(<Test path="c" state={obj} />)
  expect(selector).toBeCalledTimes(2)
  expect(selector.mock.calls[1][0]).toEqual({ x: 10, y: 3 })
  expect(selector).toHaveReturnedWith(13)
  snapshot()

  const obj2 = { ...obj }
  rerender(<Test path="c" state={obj2} />)
  expect(selector).toBeCalledTimes(2)
  snapshot()

  const obj3 = { ...obj2, b: 6 }
  rerender(<Test path="c" state={obj3} />)
  expect(selector).toBeCalledTimes(3)
  expect(selector.mock.calls[2][0]).toEqual({ x: 10, y: 6 })
  expect(selector).toHaveReturnedWith(16)
  snapshot()
})
