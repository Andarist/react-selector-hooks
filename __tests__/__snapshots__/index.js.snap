// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSelector 1`] = `
<span>
  5
</span>
`;

exports[`createSelector 2`] = `
<span>
  5
</span>
`;

exports[`createSelector 3`] = `
<span>
  7
</span>
`;

exports[`createStateSelector 1`] = `
<span>
  5
</span>
`;

exports[`createStateSelector 2`] = `
<span>
  5
</span>
`;

exports[`createStateSelector 3`] = `
<span>
  13
</span>
`;

exports[`createStateSelector 4`] = `
<span>
  13
</span>
`;

exports[`createStateSelector 5`] = `
<span>
  16
</span>
`;

exports[`createStructuredSelector 1`] = `
<span>
  5
</span>
`;

exports[`createStructuredSelector 2`] = `
<span>
  5
</span>
`;

exports[`createStructuredSelector 3`] = `
<span>
  13
</span>
`;

exports[`createStructuredSelector 4`] = `
<span>
  13
</span>
`;

exports[`createStructuredSelector 5`] = `
<span>
  16
</span>
`;
